src/vm/types/serialization.rs

```diff
@@ -1,369 +1,467 @@
-        pub fn $f_name(mut val: $type) -> String {
-            let mut result = vec![];
-            let mut negated = "";
-            
-            if val < 0 {
-                let (new_negated, new_val) = $negation(val);
-                negated = new_negated;
-                val = new_val;
-            }
 
-            loop {
-                let digit = match (val % 16) as u8 {
-                    x @  0 ..=  9 => b'0' + x,
-                    x @ 10 ..= 15 => b'a' + (x - 10),
-                    _ => panic!("number not in the range 0..15")
-                };
-                result.push(digit.into());
-                val = val / 16;
-                if val == 0 {
-            break
-                }
-            }
-            result.reverse();
-            format!("{}{}", negated, std::str::from_utf8(&result)
-                    .expect("ERROR: Hex serializer created non-utf8 string."))
+impl std::fmt::Display for SerializationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            SerializationError::IOError(e) => write!(f, "Serialization error caused by IO: {}", e.err),
+            SerializationError::BadTypeError(e) => write!(f, "Deserialization error, bad type, caused by: {}", e),
+            SerializationError::DeserializationError(e) => write!(f, "Deserialization error: {}", e),
+            SerializationError::DeserializeExpected(e) => write!(f, "Deserialization expected the type of the input to be: {}", e),
         }
     }
 }
 
-make_to_hex!(i128_to_hex, i128, |x: i128| ("-", -x));
-make_to_hex!(u128_to_hex, u128, |_x: u128| panic!("Negative UINT"));
+impl error::Error for SerializationError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            SerializationError::IOError(e) => Some(&e.err),
+            SerializationError::BadTypeError(e) => Some(e),
+            _ => None
+        }
+    }
+}
 
-fn json_simple_object(type_name: &str, val: &str) -> String {
-    format!(
-        r#"{{ "type": "{}", "value": "{}" }}"#,
-        type_name, val)
+impl From<std::io::Error> for SerializationError {
+    fn from(err: std::io::Error) -> Self {
+        match err.kind() {
+            std::io::ErrorKind::UnexpectedEof => "Unexpected end of byte stream".into(),
+            _ => SerializationError::IOError(IncomparableError { err })
+        }
+    }
 }
 
-fn json_recursive_object(type_name: &str, val: &str) -> String {
-    format!(
-        r#"{{ "type": "{}", "value": {} }}"#,
-        type_name, val)
+impl From<&str> for SerializationError {
+    fn from(e: &str) -> Self {
+        SerializationError::DeserializationError(e.into())
+    }
 }
 
-impl ClaritySerializable for Value {
-    fn serialize(&self) -> String {
+impl From<CheckErrors> for SerializationError {
+    fn from(e: CheckErrors) -> Self {
+        SerializationError::BadTypeError(e)
+    }
+}
+
+define_u8_enum!(TypePrefix {
```

Comment:
We're sure that Rust will always generate these enum u8 reprs as 0, 1, 2, ... etc.?

Reply:
`repr(u8)` is denoted as a stable representation, but we can certainly test it via some test vectors, will add.

Reply:
Added in 388ee32